Fully drop python < 3.10 support

dcbaker opened the pull request:
This fully drops python support for older platforms.

dcbaker committed:
Drop Python < 3.10 support

dcbaker committed:
ci: Drop ubuntu bionic

There is no support for Python 3.10, so we can drop this.

dcbaker committed:
ci: Pin mypy to 1.19 so we can lint Python 3.10+

mypy 1.20 fails to correctly lint python 3.10 and > 3.10, raising
warnings about unused ignore comments, when those comments are needed by
3.10

dcbaker committed:
ci/macos: use brew python for unit tests

dcbaker committed:
ci/windows: Use Python 3.10

bonzini commented on mesonbuild/mesondata.py:
```diff
@@ -25,16 +24,13 @@ def write_once(self, path: Path) -> None:
-                # fall through to python 3.7 compatible code
-                pass
+        try:
+            resource = importlib.resources.files('mesonbuild') / self.path
```
Looks like this is still needed.

dcbaker replied:
Yeah, It appears that python 3.10 isn't supported by mypy 1.20, so we need to pin back to 1.19

dcbaker commented:
I've pulled the CI image builder fixes into https://github.com/mesonbuild/meson/pull/15711, as some of these jobs will continue to fail (cross-only-armhf, for example) until the images are rebuilt

dcbaker committed:
ci: use ubuntu-rolling for non-cuda cross testing

bonzini commented:
Based on some experimentation in #15734:
* [Commit 50730ecdb8c](https://github.com/mesonbuild/meson/pull/15734/changes/50730ecdb8c31eab239994a19b6cdf265d5e0df6) "solves" the macos issue.
* #15667 correctly skips the GNOME tests in cygwin but there is another failure left in `vala/7 shared library`
* The windows/vc2022x64vs failure [is either transient or fixed by rebasing?](https://github.com/mesonbuild/meson/actions/runs/24823142788/job/72652408359?pr=15734) I didn't do anything to fix it.

Feel free to pick anything from that PR.

dcbaker commented:
I was trying to figure out whether the macos failure was because of python2 or because bare "python" was falling back to the system installed Python 3.9 instead of the brew installed 3.10+ I figured I'd wait on 15667 and rebase on that for the cygwin issues.

that very long command line test seems to be a general source of flakes. Maybe it's worth marking it as a "try 2 times"

dcbaker committed:
uninttests: Skip python2 tests if the python2 binary isn't installed

This provides faster skipping on macos and on Linux before bother to
call into pkg-config. This is needed on macOS to keep things running
after the next patch that is forced to take a heavy handed approach to
prevent the wrong python from being used.

bonzini commented:
I think your `rm` is removing a dependency so `shutil.which` doesn't catch it.

dcbaker commented:
At the moment brew is flaking out so I haven't actually seen if it works. I guess we'll find out if I can get the job to run

dcbaker commented:
seems like `shutil.which` is effective.

bonzini commented:
Should work if rebased on top of #15667!

dcbaker commented:
Finally! yay!

jon-turney commented:
I suggest there should be a release note snippet mentioning this, similar to the one that appeared in 0.62.0 (when the minimum python was raised to 3.7).

bonzini commented:
I added the needs-relnotes label so that there's no urgency to do it but it won't be forgotten.
